### DISCUSSION TITLE
Is AutoFixture maintained?

### DISCUSSION THREAD

**Discussion by aivascu** [AutoFixture/AutoFixture#1469] [General]
From time to time there are questions posted on this repository asking whether the project is alive and maintained.
This is a fair question to ask considering the low frequency of updates this project has received in the last couple of years.
I'd like to answer this question explicitly, so anyone using this library can plan accordingly.

The truth is that I currently don't have neither the time, nor the energy I had a few years ago when I just became a member of this project, in a thread with a similar topic.
The specifics of my job, leave me with little time and motivation to write code outside hours.
A couple years ago I even switched jobs to have a more balanced schedule, but that did not improve things significantly.
I've been trying to fit the work on this project in between burnouts, believing that better times are around the corner, but with periods of absence from a code-base this long, most of the time was being spent on trying to remember what I was trying to do last time.

That said the support of the AutoFixture project will **not** stop for now, however the updates will likely continue to be infrequent, with bursts of work being done every 1-2 months.
I will oversee the project until the [AutoFixture v5](https://github.com/AutoFixture/AutoFixture/discussions/1327) is released, and after that, in case my personal or the project's situation has not changed, I will try to pass the project and its assets to the .NET Foundation.

The potential transfer to the .NET Foundation will not necessarily mean extended support, as the foundation itself has a deficit of maintainers, with recent discussions being around excluding inactive projects from the foundation. This means the project will likely be archived indefinitely.

To anyone that wants to contribute, fork the project, or has any questions regarding the project, you can comment on this thread, open a new thread in the discussions, or try to reach me on Discord (@aivascu). In case there will be a lot of messages, I'll share the Discord server link.

I apologize for inconvenience this situation might have caused.

Andrei

**Comment by Kralizek** [comment]
Thanks for the message.

I'd appreciate that, before putting the project to sleep, we could relax package version dependencies by removing upper limits.

**Reply by aivascu** [reply]
yes I think it would be reasonable

**Reply by bestsoftever** [reply]
Do you guys needs any help? I have been using AutoFixutre for many years, and it's my favourite dotnet library, so I'll be happy if I could help. :)
I tried to reach you @aivascu via Discord, but seems it's not possible for me.

**Reply by aivascu** [reply]
Hi @bestsoftever,
I've stopped signing in to Discord since no one has contacted me there so far.
Help is always welcome. There's lots of things to do, starting from answering questions on SO / Discussions, and writing docs to fixing bugs, and implementing new features.
What are you interested in?

**Reply by bestsoftever** [reply]
Hi @aivascu, thanks for the response. Answering in Discussions very often requires a vision and a knowledge about the project's future, so I don't think I can help with that right now.
I think fixing bugs is better for start. But I'm not sure about the priorities - some of them are very old, are these still relevant?

**Reply by aivascu** [reply]
@bestsoftever sorry for the late reply. If you're looking for something to start with check the issues marked with the [good first issue](https://github.com/AutoFixture/AutoFixture/issues?q=is%3Aissue%20state%3Aopen%20label%3A%22good%20first%20issue%22) tag. You can also check issues marked with the [bug](https://github.com/AutoFixture/AutoFixture/issues?q=is%3Aissue%20state%3Aopen%20label%3A%22good%20first%20issue%22) tag. Some of them are marked as `triage`, but I'll remove soon the tag on those that have been confirmed for development.
If you're looking for specific suggestions I recommend having a look at one of these #1097 #1339 #1340 #1341

**Reply by bestsoftever** [reply]
@aivascu – sorry for the late response. Due to personal reasons, I haven’t had much time to spend on this project over the last couple of months. But now I’m trying to get more involved – I’ve raised a couple of PRs so far. I’m also looking into other bugs, but many of them are either old issues with open PRs that seem somewhat abandoned, and I’m not sure how to proceed with them, while the others are really hard to investigate.

Besides that, I feel a bit lost here:

- I have no idea how to build version 4 locally – I’ve installed tons of ancient SDKs, but the build keeps failing with this weird error: "Unsupported log file format. Latest supported version is 14, the log file has version 25."
- It’s a fairly large project with little documentation about the internal design. It’s hard to understand why the code is organized in certain ways – for example, the overall namespace organization, how logic is split across several classes, or why certain classes that seem internal must be public, and so on.
- StyleCop, being StyleCop.

Of course, I’m not blaming you or anyone else – the people behind this library have done truly amazing work. I just wanted to share some observations that might explain why new contributors may feel discouraged. I might be wrong, but it seems that a lack of contributors is the biggest challenge here.

**Reply by aivascu** [reply]
Hi @bestsoftever,
No worries, and thank you for the contributions. I've seen your PRs, and yesterday started reviewing them.
I'll try to get them reviewed and merged as soon as possible.

- `"Unsupported log file format. Latest supported ver...` - this seems to be an issue with the FAKE CLI. Try installing an older version, I think back when the script was written FAKE v5 was the current version. I wouldn't worry too much about v4. Your PRs are some of the last puzzle pieces that will allow the main library to be advanced to `v5`.
- `It’s a fairly large project with little documentation about the internal design.` - I tried several times to document it, but... yeah, it's a really time consuming and demotivating activity. I'll get to it once I'm done with the current release.
- `It’s hard to understand why the code is organized in certain ways` - I myself was given only an approximate description of the codebase and namespace organization, when I joined the project, and the rest was me digging through old issues, and blog posts by the original library authors.
 In my understanding, before the philosophy was that everything was public by default (see [article](https://blog.ploeh.dk/2021/03/01/pendulum-swing-internal-by-default/)), mostly because the code was written using TDD and the code had to be public in order to be tested. In the later versions I'm trying follow the internal-by-default principle, and to mimic Microsoft's approach to organizing namespaces.
- `StyleCop` - ... yeah, it's given me headaches as well, especially with the latest language changes. I think I'll remove it, it's more trouble than it's worth.

If you got specific questions feel free to ask away. I think added you on Discord, I'll try to sign into it more often. I don't really use that much lately.

**Reply by Kralizek** [reply]
Did you try have any LLM go through the code and spit out some documentation of the internals?

**Reply by aivascu** [reply]
@Kralizek not really. I tried using Copilot and Chat GPT to generate documentation on specific topics, however the models were hallucinating a lot, but I assume this is not what you're asking.
If you could show me an example of how to do it, or point me in the right direction, I could give it a try. I'm not familiar with this use case of ML, so far I only used ML only for audio transcription.

**Reply by Kralizek** [reply]
That was what I meant. Probably a problem of context size?

**Reply by aivascu** [reply]
@Kralizek don't know, could be. I'll check if there's such a setting that can be adjusted in both tools.

**Reply by bestsoftever** [reply]
Hi @aivascu,

> I've seen your PRs, and yesterday started reviewing them. I'll try to get them reviewed and merged as soon as possible.

Thanks, looking forward for some comments to address. :)
 
> * `"Unsupported log file format. Latest supported ver...` - this seems to be an issue with the FAKE CLI. Try installing an older version, I think back when the script was written FAKE v5 was the current version. I wouldn't worry too much about v4. Your PRs are some of the last puzzle pieces that will allow the main library to be advanced to `v5`.

Tried with older CLI, unfortunately same error is occurring.

> * `It’s a fairly large project with little documentation about the internal design.` - I tried several times to document it, but... yeah, it's a really time consuming and demotivating activity. I'll get to it once I'm done with the current release.

I can imagine, nobody wants to write docs. On the other hand, project without docs are not easy to use and maintain. 

> * `It’s hard to understand why the code is organized in certain ways` - I myself was given only an approximate description of the codebase and namespace organization, when I joined the project, and the rest was me digging through old issues, and blog posts by the original library authors.

Wouldn't be nice to improve this in v5?
If the code was better organized, it would be easier to write the documentation. (E.g. one chapter of the docs about a single namespace.) On the other hand, if the code was better organized, potentially less documentation would be needed.

> In my understanding, before the philosophy was that everything was public by default (see [article](https://blog.ploeh.dk/2021/03/01/pendulum-swing-internal-by-default/)), mostly because the code was written using TDD and the code had to be public in order to be tested. 

Sure, but tests shouldn't force to expose every class as public. There are at least 3 ways to address this:
- test internals by creating more test cases for public classes;
- use `InternalsVisibleTo` attribute;
- extract internal components to a separate library, to avoid polluting the API of the main library with internal details;

> * `StyleCop` - ... yeah, it's given me headaches as well, especially with the latest language changes. I think I'll remove it, it's more trouble than it's worth.

That would be great!

> If you got specific questions feel free to ask away. I think added you on Discord, I'll try to sign into it more often. I don't really use that much lately.

Thanks for the information, will try to reach you there.

**Comment by holbizmetrics** [comment]
Compatibility for NUnit 4.3.2
---
I'd basically have kind of a similar question, so I hope it is ok to post it here:

Is there a plan for AutoFixture, AutoFixture.AutoMoq, Moq to be compatible with the newest NUnit 4.3.2?

Because if I want to use NUnit 4.3.2, basically. Including all the references for unit testing in a version as is below referring to NUnit, AutoFixture, AutoFixture.AutoMoq etc.

Warning as Error occurs
---
Then this gives me:


Downgrade warning
---
```
Warning As Error: Detected package downgrade: NUnit from 4.3.2 to 3.14.0. Reference the package directly from the project to select a different version. 
 KomaxUI.UnitTests.TestCore -> AutoFixture.Nunit4 -> NUnit (>= 4.3.2) 
 KomaxUI.UnitTests.TestCore -> NUnit (>= 3.14.0)

```

Or would this be already solved by what @Kralizek mentioned:

_Thanks for the message.

I'd appreciate that, before putting the project to sleep, we could relax package version dependencies by removing upper limits._

*.csproj file example
---
This is the *.csproj file where it would downgrade as shown above:

```
<Project Sdk="Microsoft.NET.Sdk">

	<PropertyGroup>
		<TargetFramework>net8.0-windows</TargetFramework>
		<ImplicitUsings>enable</ImplicitUsings>

		<IsPackable>false</IsPackable>
		<IsTestProject>true</IsTestProject>
		<UseWPF>true</UseWPF>
	</PropertyGroup>

	<ItemGroup>
		<PackageReference Include="coverlet.collector" Version="6.0.0" />
		<PackageReference Include="Microsoft.Diagnostics.Tracing.TraceEvent" Version="3.1.19" />
		<PackageReference Include="Microsoft.NET.Test.Sdk" Version="17.8.0" />
		<PackageReference Include="Newtonsoft.Json" Version="13.0.3" />
		<PackageReference Include="NUnit" Version="3.14.0" />
		<PackageReference Include="NUnit.Analyzers" Version="3.9.0" />
		<PackageReference Include="NUnit.Engine" Version="3.18.1" />
		<PackageReference Include="NUnit3TestAdapter" Version="4.5.0" />
		<PackageReference Include="Prism.Unity" Version="8.1.97" />
		<PackageReference Include="SemanticComparison" Version="4.1.0" />
		<PackageReference Include="Unity" Version="5.11.10" />
	</ItemGroup>

	<ItemGroup>	 
		...
	</ItemGroup>

  <ItemGroup>
    <None Include="..\..\App.config" Link="testhost.dll.config">
    <CopyToOutputDirectory>Always</CopyToOutputDirectory>
    </None>
  </ItemGroup>

</Project>
```

**Reply by Kralizek** [reply]
A glue library for nunit 4 is in the works.

**Reply by aivascu** [reply]
@holbizmetrics I think you might be using the `AutoFixture.NUnit3` package instead of [AutoFixture.NUnit4](https://www.nuget.org/packages/AutoFixture.NUnit4/5.0.0-preview0012) package. As already mentioned it is still a preview but should be a full release some soon.

That said, if you experience any bugs open an issue and tag me in the description. If I don't reply within a couple days feel free to tag me again, as I might miss some notifications.

**Reply by holbizmetrics** [reply]
Thanks for the heads-up, @aivascu! The issue arose because NuGet’s dependency resolution automatically downgraded our NUnit package to version 3.x. This happened due to conflicting version constraints elsewhere in the project.

**Comment by honnes** [comment]
Hello,

Could you please provide an update about the v5 roadmap and a possible next release? I see several people working hard on this great project, and I'm a bit anxious of installing a pre-release released over a year ago, and v 4.xx of 2023 feels a bit outdated.

I just started to try out TUnit and in my search for auto-data-generation I came a cross this.

Thanks for your time and response in advance! :)

**Reply by killergege** [reply]
We use the v4 (with NUnit) extensively on a lot of projects, with thousands of tests and complicated mocks, using the greatest and latest of .NET Framework without issue.
So you can use it for new projects ! (although I didn't try it with TUnit)
